[DEVX-1047] saucectl init doc + sharding disclaimer (#504)
saucelabs/sauce-docs: docs/testrunner-toolkit/saucectl.md

See [Installation and Setup](/testrunner-toolkit/installation) for installation options.

## `$ saucectl init [flags]`

A bootstrapping command to generate a configuration file for the framework of your choice in the format required by `saucectl`. Running the command with no parameters initiates interactive mode, prompting you for relevant property values iteratively. Alternatively, you can add supported parameters inline to specify the relevant property values.

In either case, the command generates a `.sauce/config.yml` folder and file in the location from which the command is run. If you have an existing project directory for your framework, it is advised that you run this command from the project root.

```bash title="Interactive Example"
$ saucectl init
12:13:20 INF Start Init Command
? Select region: us-west-1
? Select framework:  [Use arrows to move, type to filter]
> cypress
 playwright
 puppeteer
 testcafe
 espresso
 xcuitest
```

```bash title="Batch Example"

> $saucectl init -r us-west-1 -f cypress -b chrome

Reviewer: Slight readability enhancement

```suggestion
$ saucectl init -r us-west-1 -f cypress -b chrome
```

```

`saucectl` supports the following configuration flags as inline specifications.

> ### `-a, --accessKey`

Reviewer: Question opened for every field:
Shouldn't we say in that title that there is a parameter ?
In `saucectl init` case all have one but `--help`, and at first glance it may be not clear that a value is needed.

Was thinking of something like that:
```suggestion
### `-a <string>, --accessKey <string>`
```

<p><small>| OPTIONAL | STRING |</small></p>

The authentication access key associated with the Sauce Labs user account making this request. If you have not set your authentication credentials as environment parameters or generated a `credentials.yml` file, this value is required.

---

> ### `-app`

Reviewer: ```suggestion
### `--app`
```

<p><small>| OPTIONAL | STRING | XCUITEST/ESPRESSO ONLY |</small></p>

The path to a valid mobile application to test.

---

### `--artifacts.download.when`

<p><small>| OPTIONAL | STRING |</small></p>

Specifies when and under what circumstances to download artifacts. Valid values are:

* `always`: Always download artifacts.
* `never`: Never download artifacts.
* `pass`: Download artifacts for passing suites only.
* `fail`: Download artifacts for failed suites only. (default value)

---

### `-b, --browserName`
<p><small>| REQUIRED | STRING | WEB-APP ONLY |</small></p>

The name of the browser in which to run tests.

---

### `--cypress.config`
<p><small>| REQUIRED | STRING | CYPRESS ONLY |</small></p>

The file path to the Cypress configuration file (typically `cypress.json`).

---

### `--device`
<p><small>| OPTIONAL | STRING | XCUITEST/ESPRESSO ONLY |</small></p>

Find a real device for this test matching the device name or portion of the name provided. Use the [Get Devices API Request](https://docs.saucelabs.com/dev/api/rdc#get-devices) for a list of supported real devices in your region.

> 

Reviewer: In that specific case, we should provide an example, it's not as simple as specifying the device name.
For consistency, it's similar to the cli-driven approach for espresso.

There is different flags: 
```
id
name
platformName
platformVersion
carrierConnectivity
deviceType
private
```

Examples:
- Just filtering by name:
  ```
  --device "name=HTC.*"
  ```
- Several flags can be specified while being comma separated
   ```
  --device "name=HTC.*,platformVersion=8.0.0"
   ```

Author: OH!! That does change things. LOL I will make the clarification. Thanks for this.

Author: @FriggaHel -- are you sure the `platformName` is one of the supported flags for device? It is not one of the devices properties in the config file.

---

### `--emulator`
<p><small>| OPTIONAL | STRING | ESPRESSO ONLY |</small></p>

The name of the virtual device emulator for the test. Check the list of [supported virtual devices](https://app.saucelabs.com/live/web-testing/virtual) for valid names.

> 

Reviewer: Similar to `--device`.
Available fields are:

```
name
orientation
platformVersion
```

Example:
```
--emulator 'name=Samsung Galaxy S8 FHD GoogleAPI Emulator,platformVersion=7.1'
```

---

### `-f, --framework`
<p><small>| REQUIRED | STRING |</small></p>

The framework for which this configuration is intended.

---

### `-v, --frameworkVersion`
<p><small>| REQUIRED | STRING | WEB APPS ONLY |</small></p>

The version of the framework that is compatible with the tests defined in this configuration.

---

### `-h, --help`

Usage information for the `init` command.

---

### `-p, --platformName`

> <p><small>| OPTIONAL | STRING |</small></p>

Reviewer: To be consistent with `--framework` and other flags.

```suggestion
<p><small>| OPTIONAL | STRING | WEB APPS ONLY |</small></p>
```

Platform is implicit for XCUITest & Espresso.

> A specific operating system and version on which to run the specified browser and test suite. Defaults to a platform that is supported by `saucectl` for the specified browser.

Reviewer: There is no default value to that one.
Notice that this is the place where you can specifiy `docker` as a platform.

Specify an alternative configuration file to the default `.sauce/config.yml` for this execution.

```bash
saucectl run -c ./path/to/{config-file}.yml
```

:::note YAML Required
While you can use multiple files of different names or locations to specify your configurations, each file must be a `*.yml` and follow the `saucectl` syntax. If you are less comfortable with YAML, any of a wide variety of free online YAML/JSON validator tools may be helpful.
:::
---

### `--env`
<p><small>| OPTIONAL | KEY=VALUE |</small></p>

An environment variable key value pair that may be referenced in the tests executed by this command. Expanded environment variables are supported.

```bash
saucectl run --env <key>=value1> --env <key2>=<value2> ...
```
---

### `-r, --region`
<p><small>| REQUIRED | STRING |</small></p>

Specifies the Sauce Labs data center through which tests will run. Valid values are: `us-west-1` (default) or `eu-central-1`.

```bash
saucectl run --region use-west-1
```
---

### `--suite`
<p><small>| OPTIONAL | STRING |</small></p>

Specifies a test suite to execute by name.

```bash
saucectl run --suite <suite_name>
```
---

### `--tags`
<p><small>| OPTIONAL | LIST |</small></p>

> A keyword that may help you distinguish the test in Sauce Labs, and also helps you apply filters to easily isolate tests based on metrics that are meaningful to you

Reviewer: ```suggestion
A keyword that may help you distinguish the test in Sauce Labs, and also helps you apply filters to easily isolate tests based on metrics that are meaningful to you.
```

<p><small>| OPTIONAL | STRING |</small></p>

The authentication access key associated with the Sauce Labs user account making this request. If you have not set your authentication credentials as environment parameters or generated a `credentials.yml` file, this value is required.

**Shorthand:** `-a <string>`

---

### `--app <string>`
<p><small>| OPTIONAL | STRING | XCUITEST/ESPRESSO ONLY |</small></p>

The path to a valid mobile application to test.

---

### `--artifacts.download.when <string>`

<p><small>| OPTIONAL | STRING |</small></p>

Specifies when and under what circumstances to download artifacts. Valid values are:

* `always`: Always download artifacts.
* `never`: Never download artifacts.
* `pass`: Download artifacts for passing suites only.
* `fail`: Download artifacts for failed suites only. (default value)

---

### `--browserName <string>`
<p><small>| REQUIRED | STRING | WEB-APP ONLY |</small></p>

The name of the browser in which to run tests.

**Shorthand:** `-b <string>`

---

### `--cypress.config <string>`
<p><small>| REQUIRED | STRING | CYPRESS ONLY |</small></p>

The file path to the Cypress configuration file (typically `cypress.json`).

---

> ### `--device ["flag=value"]`
> <p><small>| OPTIONAL | ARRAY | XCUITEST/ESPRESSO ONLY |</small></p>

Reviewer: ```suggestion
### `--device <string>`
<p><small>| OPTIONAL | STRING | XCUITEST/ESPRESSO ONLY |</small></p>
```

IMO, that's better this way.
This is not an array per-se, just a big long string depending the number of flags.

`[]` in cli helps oftenly denotes an optional parameter, whereas `<>` denotes a mandatory one.

Here is a more in-depth explanations: https://ftpdocs.broadcom.com/cadocs/0/CA%20ARCserve%20%20Backup%2015-ENU/Bookshelf_Files/HTML/CMD_Ref/command_line_syntax_characters.htm

Find a real device for this test by matching a set of one or more device characteristics:

|Characteristic|Description|Example|
|---|---|---|
|`id`| Specify a device by its ID. Using this selection flag ignores all other characteristics and is not advised because availability of a specific device is uncertain and could cause your test to time out.| ```--device "id=HTC_U11_real_us"```|
|`name`|Find a device based on a partial name in order to increase likelihood of availability of similar devices.|```--device "name=HTC.*"```|
|`platformVersion`|Find a device based on its platform version.|```--device "platformVersion=8.0"```|
|`carrierConnectivity`|The selected device must be connected to a cellular network. |```--device "carrierConnectivity=true"```|
|`deviceType`|The selected device must be a particular type (`PHONE`, `TABLET`, or `ANY`).|```--device "deviceType=PHONE"```|
|`private`|The selected device must be private.|```--device "private=true"```|

You can specify a combination of device characteristics within this flag:

```bash
--device "name=HTC.*,platformVersion=8.0.0,carrierConnectivity=true"
```

---

> ### `--emulator ["flag=value"]`

Reviewer: ```suggestion
### `--emulator <string>`
```

Same as above

**Shorthand:** `-v <string>`

---

### `-h, --help`

Usage information for the `init` command.

---

### `--platformName <string>`
<p><small>| OPTIONAL | STRING | WEB APPS ONLY |</small></p>

A specific operating system and version on which to run the specified browser and test suite.

:::note
You can optionally specify `docker` here as the platform.
:::

**Shorthand:** `-p <string>`

---

### `--region <string>`
<p><small>| REQUIRED | STRING |</small></p>

Specifies the Sauce Labs data center through which tests will run. Valid values are: `us-west-1` (default) or `eu-central-1`.

**Shorthand:** `-r <string>`

---

### `--testApp <string>`
<p><small>| REQUIRED | STRING | XCUITEST/ESPRESSO ONLY |</small></p>

The path to the mobile testing application.

**Shorthand:** `-t <string>`

---

### `--username <string>`
<p><small>| OPTIONAL | STRING |</small></p>

A valid Sauce Labs user account. If you have not set your authentication credentials as environment parameters or generated a `credentials.yml` file, this value is required.

**Shorthand:** `-u <string>`

---

## `$ saucectl configure [flags]`

> Allows you to provide your [Sauce Labs credentials]((https://app.saucelabs.com/user-settings) for the purpose of generating a `credentials.yml` file that `saucectl` can access to automatically authenticate commands without requiring manual authentication. The `credentials.yml` file is created in a `'sauce` folder in your `saucectl` installation root.

Reviewer: ```suggestion
Allows you to provide your [Sauce Labs credentials]((https://app.saucelabs.com/user-settings) for the purpose of generating a `credentials.yml` file that `saucectl` can access to automatically authenticate commands without requiring manual authentication. The `credentials.yml` file is created in a `.sauce` folder in your  home directory.
```

<p><small>| OPTIONAL | STRING |</small></p>

Associates the tests with a build.

```bash
saucectl run --build myBuildID
```
---

### `--concurrency <int>`
<p><small>| OPTIONAL | INTEGER |</small></p>

Sets the maximum number of suites to execute at the same time. If the test defines more suites than the max, excess suites are queued and run in order as each suite completes.

**Shorthand:** `-ccy`

:::caution
For tests running on Sauce, set this value to equal or less than your Sauce concurrency allowance, as setting a higher value may result in jobs dropped by the server.
:::

```bash
saucectl run --ccy 2
```
---

### `--config <path>`
<p><small>| OPTIONAL | FILEPATH |</small></p>

Specify an alternative configuration file to the default `.sauce/config.yml` for this execution.

**Shorthand:** `-c <path>`

```bash
saucectl run -c ./path/to/{config-file}.yml
```

:::note YAML Required
While you can use multiple files of different names or locations to specify your configurations, each file must be a `*.yml` and follow the `saucectl` syntax. If you are less comfortable with YAML, any of a wide variety of free online YAML/JSON validator tools may be helpful.
:::
---

> ### `--env [key=value]`

Reviewer: ```suggestion
### `--env <string>`
```
Same as above

<p><small>| OPTIONAL | KEY=VALUE |</small></p>

An environment variable key value pair that may be referenced in the tests executed by this command. Expanded environment variables are supported.

```bash
saucectl run --env <key1>=value1> --env <key2>=<value2> ...
```
---

### `--region <string>`
<p><small>| REQUIRED | STRING |</small></p>

Specifies the Sauce Labs data center through which tests will run. Valid values are: `us-west-1` (default) or `eu-central-1`.

**Shorthand:** `-r <string>`

```bash
saucectl run --region use-west-1
```
---

### `--suite <string>`
<p><small>| OPTIONAL | STRING |</small></p>

Specifies a test suite to execute by name.

```bash
saucectl run --suite <suite_name>
```
---

> ### `--tags [list]`

Reviewer: ```suggestion
### `--tags <tag1,tag2,...>`
```
